apps/game_backend/lib/game_backend/users.ex

```diff
@@ -113,4 +130,56 @@ defmodule GameBackend.Users do
     })
     |> Repo.update()
   end
+
+  defp apply_currency_rewards(user_id, currency_rewards) do
+    Enum.map(currency_rewards, fn currency_reward ->
+      Currencies.add_currency(user_id, currency_reward.currency_id, currency_reward.amount)
+    end)
+    |> check_result(:currency_rewards)
+  end
+
+  defp build_item_rewards_params(user_id, item_rewards) do
+    Enum.map(item_rewards, fn item_reward ->
+      Enum.map(1..item_reward.amount, fn _ ->
+        %{
+          user_id: user_id,
+          template_id: item_reward.item_template_id,
+          level: 1,
+          inserted_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second),
+          updated_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
+        }
+      end)
+    end)
+    |> List.flatten()
+  end
+
+  defp build_unit_rewards_params(user_id, unit_rewards) do
+    IO.inspect(unit_rewards, label: :rewards)
+
+    Enum.map(unit_rewards, fn unit_reward ->
+      Enum.map(1..unit_reward.amount, fn _ ->
+        %{
+          user_id: user_id,
+          character_id: unit_reward.character_id,
+          unit_level: 1,
+          tier: 1,
+          selected: false,
+          inserted_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second),
+          updated_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
+        }
+      end)
+    end)
+    |> List.flatten()
+  end
+
+  defp check_result(result, element_name) do
+    if Enum.all?(result, fn
+         {:ok, _} -> true
+         _ -> false
+       end) do
+      {:ok, result}
+    else
+      {:error, "Failed to apply " <> Atom.to_string(element_name)}
+    end
+  end
```

Comment:
a few things about this:
1. You're just using this function for `apply_currency_rewards` and imo the can be in the same function
2. the enum all clause is a bit weird, why not just comparing the first element of the tuple? 
```
if Enum.all?(result, fn {result, _} -> result == :ok end)
```
3. this may be a personal opinion but why not adding the currency updating to the main multi? that way you don't need this validation
```
      Multi.new()
      |> apply_currency_rewards(user_id, level.currency_rewards)



defp apply_currency_rewards(multi, user_id, currency_rewards) do
  Enum.reduce(currency_rewards, multi, fn currency_reward, multi -> 
    Multi.run({:add_currency, currency_reward}, Currencies.add_currency(user_id, currency_reward.currency_id, currency_reward.amount))
  end)
end
```